feat(superchain): add GovernedByOptimism field to chain config

edobry opened the pull request:
### Description

This PR adds a boolean `GovernedByOptimism` field to the chain config, which indicates if a certain chain is subject to Optimism governance. It also adds a `GovernedByOptimismTest` test to the `validation` package, which ensures that any chain which has the `governed_by_optimism` field set to `true` in its config also has its `ProxyAdmin` address set to the Optimism multisig.

It also sets the new `governed_by_optimism` field on all chain configs as appropriate

### Metadata

- **In service of**: #676 
- **Depends on:** #717

geoknee commented:
Two main questions:

1. If we declare a chain as governed by optimism, but it actually isn't, how would we tell?

I am wondering if this change be accompanied by some edits to this validation check:

https://github.com/ethereum-optimism/superchain-registry/blob/c56b99231b9033d4db7d64f4fe4eeca72d733a8a/validation/key-handover_test.go#L12-L32

EDIT: just spotted the other PR #719. I think it would make sense to commit the changes in a single PR, if possible.

2. When does this boolean get written / updated? During the `add-chain` tool? Manually? During promotion to a standard chain? Do we need to update the .env schema / add-chain tool at all?

bitwiseguy commented on validation/governed-by-optimism_test.go:
```diff
@@ -0,0 +1,23 @@
+	"github.com/stretchr/testify/require"
+)
+
+func testGovernedByOptimism(t *testing.T, chain *ChainConfig) {
+	chainID := chain.ChainID
+	superchain := OPChains[chainID].Superchain
+
+	optimismMultisig := standard.Config.MultisigRoles[superchain].KeyHandover.L1.Universal["ProxyAdmin"]["owner()"]
+
+	if chain.GovernedByOptimism == true {
+		require.Equal(t, optimismMultisig, chain.Addresses.ProxyAdminOwner.String(), "Chains using Optimism governance must have their ProxyAdminOwner set to the Optimism multisig")
+	}
+}
```
Could this be combined with the existing `key-handover_test.go` code? Seems like that test is doing basically the same thing, except for checking the new boolean

edobry commented on superchain/configs/mainnet/op.toml:
```diff
@@ -4,6 +4,7 @@ public_rpc = "https://mainnet.optimism.io"
 sequencer_rpc = "https://mainnet-sequencer.optimism.io"
 explorer = "https://explorer.optimism.io"
 superchain_level = 1
+governed_by_optimism = true
```
I was thinking that it may not be important to set it for the `false` cases, but I can go ahead and do this

bitwiseguy commented on validation/governed-by-optimism_test.go:
```diff
@@ -0,0 +1,22 @@
+	. "github.com/ethereum-optimism/superchain-registry/superchain"
+	"github.com/ethereum-optimism/superchain-registry/validation/standard"
+
+	// "github.com/ethereum/go-ethereum/common"
```
This commented line can be removed

bitwiseguy commented on validation/governed-by-optimism_test.go:
```diff
@@ -0,0 +1,22 @@
+	chainID := chain.ChainID
+	superchain := OPChains[chainID].Superchain
+
+	if chain.GovernedByOptimism == true {
+		optimismMultisig := standard.Config.MultisigRoles[superchain].KeyHandover.L1.Universal["ProxyAdmin"]["owner()"]
+		require.Equal(t, optimismMultisig, chain.Addresses.ProxyAdminOwner.String(), "Chains using Optimism governance must have their ProxyAdminOwner set to the Optimism multisig")
```
Should we also check for `chain.Addresses.ProxyAdminOwner.String() != optimismMultisig` in the case that `chain.GovernedByOptimism == false`? It would be nice if we did that so that this validation check would fail if someone updated a chain config's `ProxyAdminOwner` but forgets to update the `governed_by_optimism` bool

edobry replied:
done

tessr commented:
Just chiming in to confirm that this can be the same logic as the key handover test. We are functionally checking the same thing (whether the proxy admin owner is set to the correct address).

opgitgovernance commented:
Detected nonzero diff in rollup config. This PR modifies the configuration of at least one chain in the registry, which may or may not have been intended. You can see the detail [here](https://circleci.com/gh/ethereum-optimism/superchain-registry/50281).

opgitgovernance commented:
Detected nonzero diff in genesis. This PR modifies the genesis data of at least one chain in the registry, which may or may not have been intended. You can see the detail [here](https://circleci.com/gh/ethereum-optimism/superchain-registry/50280).

edobry commented:
I've updated this PR with several changes:

1. restructured the `validation/standard/*.toml` files to improve DevX
2. added the negative test case [suggested here](https://github.com/ethereum-optimism/superchain-registry/pull/718#discussion_r1869691265)
3. added standard configs for the `sepolia-dev-0` chain to account for newly-failing tests caused by 2.

I explored whether we can reuse the `Key_Handover` test for this usecase but this doesn't seem possible, as it doesn't check the configured address directly but rather calls a contract method; I tested this by making changes in the config file which should've caused it to fail, but it didn't, not fully sure why. Please LMK if I've misunderstood something here.

bitwiseguy commented on validation/standard/config/networks/mainnet/standard-config-roles-mainnet.toml:
I believe some of these `standard-*.toml` files are referenced in the governance blockspace charter. @tessr @geoknee do you know if moving these files around will break any links or cause any other issues? If so, I would suggest reverting these toml file refactors from within this pr. Some of the proposed op-deployer/superchain-registry integration work will refactor these files as well, so might be best to postpone this part of the work and just do it once

bitwiseguy commented on validation/security-configs_test.go:
```diff
@@ -17,29 +17,42 @@ import (
 
-var checkResolutions = func(t *testing.T, r standard.Resolutions, chainID uint64, client *ethclient.Client) {
-	for contract, methodToOutput := range r {
+func resolveAddress(t *testing.T, chainID uint64, contract string) (Address, error) {
```
Nit: could we change this function name to `getAddressFromConfig` and change the argument `contract` --> `contractName`

bitwiseguy commented on validation/standard/config/networks/sepolia-dev-0/standard-versions-sepolia-dev-0.toml:
Have you verified that the addresses in this file are correct? We currently have a gap in our validations that doesn't actually check that the addresses used by a chain match the addresses in this file, which is likely the reason we aren't seeing an error for sepolia-dev-0 now that this file exists

edobry replied:
changes moved to https://github.com/ethereum-optimism/superchain-registry/pull/761

opgitgovernance commented:
Detected nonzero diff in rollup config. This PR modifies the configuration of at least one chain in the registry, which may or may not have been intended. You can see the detail [here](https://circleci.com/gh/ethereum-optimism/superchain-registry/51847).

opgitgovernance commented:
Detected nonzero diff in genesis. This PR modifies the genesis data of at least one chain in the registry, which may or may not have been intended. You can see the detail [here](https://circleci.com/gh/ethereum-optimism/superchain-registry/51846).

bitwiseguy commented on validation/standard/init_test.go:
```diff
@@ -28,18 +28,28 @@ func TestConfigInitialization(t *testing.T) {
-			require.NotNil(t, Config.MultisigRoles[network], "Config.MultisigRoles[%s] should not be nil", network)
-			require.NotZero(t, Config.MultisigRoles[network], "Config.MultisigRoles[%s] should not be zero value", network)
+			require.NotNil(t, roles, "Config.MultisigRoles[%s] should not be nil", network)
+			if roles != nil {
```
Nit: since you require that `roles != nil` on the line above, don't think you need this `if` statement. Same for the `if l1Roles != nil` statement below

edobry replied:
my thinking was that since `require.NotNil` only fails the test, the program would continue and could potentially panic on the next line, and that could confuse the test output

bitwiseguy replied:
I think `require.NotNil` will exit the test execution immediately upon failure. This is different from `assert.NotNil`, which behaves the way you describe

edobry replied:
nvm tested again and you're right, i'll make this change

bitwiseguy reviewed (approved):
LGTM
